README.md
# scaffold-rule

> support  `eslint prettier commitlint stylelint`

## eslint

> JSDOC

## prettier

## commitlint

## stylelint


CHANGELOG.md
<a name="0.5.0"></a>
# 0.5.0 (2021-02-18)


### Bug Fixes

* lock ([601ac58](https://github.com/Objectivezt/scaffold-rule/commit/601ac58))
* remove package-lock.json ([9c9408d](https://github.com/Objectivezt/scaffold-rule/commit/9c9408d))


### Features

*     complexity: [ ([87c22cc](https://github.com/Objectivezt/scaffold-rule/commit/87c22cc))
* add eslint prettier config ([b3a3956](https://github.com/Objectivezt/scaffold-rule/commit/b3a3956))
* add htmlhint jsdoc lint ([8c2314d](https://github.com/Objectivezt/scaffold-rule/commit/8c2314d))
* add lintstage ([d0535e1](https://github.com/Objectivezt/scaffold-rule/commit/d0535e1))
* add stylelintrc ([3649bec](https://github.com/Objectivezt/scaffold-rule/commit/3649bec))
* changelog auto stylelint css config update ([672c360](https://github.com/Objectivezt/scaffold-rule/commit/672c360))
* gitignore update ([00b26c4](https://github.com/Objectivezt/scaffold-rule/commit/00b26c4))
* rename package name ([284a8ed](https://github.com/Objectivezt/scaffold-rule/commit/284a8ed))





.prettierrc.js
/*
 * @Author: objectivezt
 * @Date: 2021-02-16 17:13:30
 * @Last Modified by: objectivezt
 * @Last Modified time: 2021-02-17 14:37:24
 */

module.exports = {
  /**
   * @english Use single quotes instead of double quotes in JSX.
   * @chinese 在JSX中使用单引号而不是双引号
   */
  singleQuote: true,
  /**
   * @english Put the > of a multi-line JSX element at the end of the last line instead of being alone on the next line (does not apply to self closing elements).
   * @chinese 将多行JSX元素的>放在最后一行的末尾，而不是单独放在下一行（不适用于自闭元素）
   */
  jsxBracketSameLine: true,
  /**
   * @english endOfLine: "<lf|crlf|cr|auto>
   * @chinese
   */
  endOfLine: 'auto',

  printWidth: 100, // 120
  /**
   * @english By default, Prettier will wrap markdown text as-is since some services use a linebreak-sensitive renderer, e.g. GitHub comment and BitBucket. In some cases you may want to rely on editor/viewer soft wrapping instead, so this option allows you to opt out with "never".
   * @chinese
   */
  proseWarp: 'never',
};


.stylelintrc.js
/*
 * @Author: objectivezt
 * @Date: 2021-02-17 10:35:39
 * @Last Modified by: objectivezt
 * @Last Modified time: 2021-02-17 15:10:06
 */
module.exports = {
  extends: [
    'stylelint-config-prettier',
    'stylelint-config-standard',
    'stylelint-config-recess-order',
    './node_modules/prettier-stylelint/config.js'
  ],
  plugins: ['stylelint-selector-bem-pattern', 'stylelint-order'],
  rules: {
    'plugin/selector-bem-pattern': {
      // 选择Preset Patterns，支持suit和bem两种，默认suit规范；
      // 不管哪种都需要手动指定，因为该插件未给源插件默认指定
      preset: 'bem',
      /**
       * 自定义模式规则
       * 指定组合的选择器检查规则，其实就是指定class名称规则
       * 支持正则字符串、返回正则的函数、包含2个选项参数的对象等写法
       */
      componentSelectors: {
        // 只初始的选择器规则（可以理解为外层的class规则）
        initial: '^\\.{componentName}(?:__[-a-z]+)?(?:--[a-z]+)?$',
        // 可以理解为外层以内的选择器的规则，
        // 如果不指定，则跟initial同样的规则，
        // 注意这里配置的时候比上面少一个问号，
        // 是希望内层就不应该只有componentName作为选择器了
        combined: '^\\.{componentName}(?:__[-a-z]+)(?:--[a-z]+)?$'
      },
      utilitySelectors: '^\\.u-[a-z]+$',
      ignoreSelectors: ['^\\.el-', '/deep/', '>>>', '^\\.icon-'],
      ignoreCustomProperties: []
    },
    'order/properties-order': [
      {
        // Must be first.
        properties: ['all']
      },
      {
        // Position.
        properties: ['position', 'top', 'right', 'bottom', 'left', 'z-index']
      },
      {
        // Display mode.
        properties: ['box-sizing', 'display']
      },
      {
        // Flexible boxes.
        properties: [
          'flex',
          'flex-basis',
          'flex-direction',
          'flex-flow',
          'flex-grow',
          'flex-shrink',
          'flex-wrap'
        ]
      },
      {
        // Grid layout.
        properties: [
          'grid',
          'grid-area',
          'grid-template',
          'grid-template-areas',
          'grid-template-rows',
          'grid-template-columns',
          'grid-row',
          'grid-row-start',
          'grid-row-end',
          'grid-column',
          'grid-column-start',
          'grid-column-end',
          'grid-auto-rows',
          'grid-auto-columns',
          'grid-auto-flow',
          'grid-gap',
          'grid-row-gap',
          'grid-column-gap'
        ]
      },
      {
        // Align.
        properties: ['align-content', 'align-items', 'align-self']
      },
      {
        // Justify.
        properties: ['justify-content', 'justify-items', 'justify-self']
      },
      {
        // Order.
        properties: ['order']
      },
      {
        // Box model.
        properties: [
          'float',
          'width',
          'min-width',
          'max-width',
          'height',
          'min-height',
          'max-height',
          'padding',
          'padding-top',
          'padding-right',
          'padding-bottom',
          'padding-left',
          'margin',
          'margin-top',
          'margin-right',
          'margin-bottom',
          'margin-left',
          'overflow',
          'overflow-x',
          'overflow-y',
          '-webkit-overflow-scrolling',
          '-ms-overflow-x',
          '-ms-overflow-y',
          '-ms-overflow-style',
          'clip',
          'clear'
        ]
      },
      {
        // Typography.
        properties: [
          'font',
          'font-family',
          'font-size',
          'font-style',
          'font-weight',
          'font-variant',
          'font-size-adjust',
          'font-stretch',
          'font-effect',
          'font-emphasize',
          'font-emphasize-position',
          'font-emphasize-style',
          '-webkit-font-smoothing',
          '-moz-osx-font-smoothing',
          'font-smooth',
          'hyphens',
          'line-height',
          'color',
          'text-align',
          'text-align-last',
          'text-emphasis',
          'text-emphasis-color',
          'text-emphasis-style',
          'text-emphasis-position',
          'text-decoration',
          'text-indent',
          'text-justify',
          'text-outline',
          '-ms-text-overflow',
          'text-overflow',
          'text-overflow-ellipsis',
          'text-overflow-mode',
          'text-shadow',
          'text-transform',
          'text-wrap',
          '-webkit-text-size-adjust',
          '-ms-text-size-adjust',
          'letter-spacing',
          'word-break',
          'word-spacing',
          'word-wrap', // Legacy name for `overflow-wrap`
          'overflow-wrap',
          'tab-size',
          'white-space',
          'vertical-align',
          'list-style',
          'list-style-position',
          'list-style-type',
          'list-style-image'
        ]
      },
      {
        // Accessibility & Interactions.
        properties: [
          'pointer-events',
          '-ms-touch-action',
          'touch-action',
          'cursor',
          'visibility',
          'zoom',
          'table-layout',
          'empty-cells',
          'caption-side',
          'border-spacing',
          'border-collapse',
          'content',
          'quotes',
          'counter-reset',
          'counter-increment',
          'resize',
          'user-select',
          'nav-index',
          'nav-up',
          'nav-right',
          'nav-down',
          'nav-left'
        ]
      },
      {
        // Background & Borders.
        properties: [
          'background',
          'background-color',
          'background-image',
          "-ms-filter:\\'progid:DXImageTransform.Microsoft.gradient",
          'filter:progid:DXImageTransform.Microsoft.gradient',
          'filter:progid:DXImageTransform.Microsoft.AlphaImageLoader',
          'filter',
          'background-repeat',
          'background-attachment',
          'background-position',
          'background-position-x',
          'background-position-y',
          'background-clip',
          'background-origin',
          'background-size',
          'background-blend-mode',
          'isolation',
          'border',
          'border-color',
          'border-style',
          'border-width',
          'border-top',
          'border-top-color',
          'border-top-style',
          'border-top-width',
          'border-right',
          'border-right-color',
          'border-right-style',
          'border-right-width',
          'border-bottom',
          'border-bottom-color',
          'border-bottom-style',
          'border-bottom-width',
          'border-left',
          'border-left-color',
          'border-left-style',
          'border-left-width',
          'border-radius',
          'border-top-left-radius',
          'border-top-right-radius',
          'border-bottom-right-radius',
          'border-bottom-left-radius',
          'border-image',
          'border-image-source',
          'border-image-slice',
          'border-image-width',
          'border-image-outset',
          'border-image-repeat',
          'outline',
          'outline-width',
          'outline-style',
          'outline-color',
          'outline-offset',
          'box-shadow',
          'mix-blend-mode',
          'filter:progid:DXImageTransform.Microsoft.Alpha(Opacity',
          "-ms-filter:\\'progid:DXImageTransform.Microsoft.Alpha",
          'opacity',
          '-ms-interpolation-mode'
        ]
      },
      {
        // SVG Presentation Attributes.
        properties: [
          'alignment-baseline',
          'baseline-shift',
          'dominant-baseline',
          'text-anchor',
          'word-spacing',
          'writing-mode',

          'fill',
          'fill-opacity',
          'fill-rule',
          'stroke',
          'stroke-dasharray',
          'stroke-dashoffset',
          'stroke-linecap',
          'stroke-linejoin',
          'stroke-miterlimit',
          'stroke-opacity',
          'stroke-width',

          'color-interpolation',
          'color-interpolation-filters',
          'color-profile',
          'color-rendering',
          'flood-color',
          'flood-opacity',
          'image-rendering',
          'lighting-color',
          'marker-start',
          'marker-mid',
          'marker-end',
          'mask',
          'shape-rendering',
          'stop-color',
          'stop-opacity'
        ]
      },
      {
        // Transitions & Animation.
        properties: [
          'transition',
          'transition-delay',
          'transition-timing-function',
          'transition-duration',
          'transition-property',
          'transform',
          'transform-origin',
          'animation',
          'animation-name',
          'animation-duration',
          'animation-play-state',
          'animation-timing-function',
          'animation-delay',
          'animation-iteration-count',
          'animation-direction'
        ]
      }
    ]
  }
};


.eslintrc.js
/*
 * @Author: objectivezt
 * @Date: 2021-02-16 17:47:42
 * @Last Modified by: objectivezt
 * @Last Modified time: 2021-02-17 14:26:29
 */
const NORMAL_MODE = {
  env: {
    browser: true,
    es6: true,
    node: true,
  },
  extends: ['eslint:recommended', 'plugin: react/recommended'],
  globals: {
    Atomics: 'readonly',
    SharedArrayBuffer: 'readonly',
  },
  settings: {
    react: {
      version: 'detect',
    },
  },
  parser: 'babel-eslint',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
      legacyDecorators: true
    },
    ecmaVersion: 2018,
    sourceType:'module'
  },
  plugins: ['react', 'prettier'],
  rules: {
    'prettier/prettier': 2,
    'react/prop-types': 0,
    'react/display-name': 0,
    'no-extra-boolean-cast': 0,
    'react/no-deprecated': 0
  }
};

const STRICT_MODE = {
  env: {
    browser: true,
    node: true,
    es6: true,
    mocha: true,
    jest: true,
    jasmine: true,
  },
  extends: [
    'airbnb',
    'prettier',
    'prettier/react',
    'plugin:prettier/recommended',
    // append in 20210216
    'airbnb-typescript',
    'prettier/@typescript-eslint',
    // append in 20210217
    'plugin:jsdoc/recommended',
  ],
  parser: 'babel-eslint',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
      legacyDecorators: true,
    },
  },
  plugins: [
    'react',
    'jest',
    'prettier',
    'jsx-a11y',
    'import',
    // append in 20210216
    '@typescript-eslint',
    'eslint-comments',
    'unicorn',
    'react-hooks',
    // append in 20210217
    'jsdoc'
  ],
  // append in 20210216
  settings: {
    // support import modules from TypeScript files in JavaScript files
    'import/resolver': { node: { extensions: ['.js', '.jsx', '.ts', '.tsx'] } },
    polyfills: ['fetch', 'Promise', 'URL', 'object-assign'],
  },
  rules: {
    // reference umi
    'array-callback-return': 0,
    'arrow-body-style': ['error', 'as-needed'],
    'arrow-parens': 0,
    'function-paren-newline': 0,
    'generator-star-spacing': 0,
    'global-require': 0,
    'implicit-arrow-linebreak': 0,
    'import/extensions': 0,
    'import/no-cycle': 0,
    'import/no-default-export': [0, 'camel-case'],
    'import/no-dynamic-require': 0,
    'import/no-extraneous-dependencies': 0,
    'import/no-unresolved': [
      2,
      {
        ignore: [
          '^@/',
          '^@assets/',
          '^@common/',
          '^@components/',
          '^@containers/',
          '^@layouts/',
          '^@models/',
          '^@services/',
          '^@styles/',
          '^@utils/',
        ],
      },
    ],
    'import/order': 1,
    'import/prefer-default-export': 0,
    'jsx-a11y/anchor-is-valid': 0,
    'jsx-a11y/click-events-have-key-events': 0,
    'jsx-a11y/no-noninteractive-element-interactions': 0,
    'jsx-a11y/no-static-element-interactions': 0,
    'linebreak-style': 0,
    'no-else-return': 0,
    'no-extra-boolean-cast': 0,
    'no-param-reassign': 1,
    'no-plusplus': 0,
    'no-prototype-builtins': 0,
    'no-useless-escape': 0,
    'no-use-before-define': ['error', { functions: false, classes: true, variables: true }],
    'object-curly-newline': 0,
    'operator-linebreak': 0,
    'prettier/prettier': 2,
    'react/destructuring-assignment': 0,
    'react/display-name': 0,
    'react/forbid-prop-types': 0,
    'react/jsx-closing-bracket-location': 0,
    'react/jsx-filename-extension': 0,
    'react/jsx-fragments': 0,
    'react/jsx-one-expression-per-line': 0,
    'react/jsx-props-no-spreading': 0,
    'react/jsx-wrap-multilines': 0,
    'react/no-deprecated': 0,
    'react/prop-types': 0,
    'react/sort-comp': 1,
    'react/static-property-placement': 0,
    'space-before-function-paren': 0,
    'unicorn/prevent-abbreviations': 0,

    // append 20200216
    '@typescript-eslint/dot-notation': 0,
    '@typescript-eslint/explicit-member-accessibility': 0,
    '@typescript-eslint/explicit-function-return-type': [
      'off',
      { allowTypedFunctionExpressions: true },
    ],
    '@typescript-eslint/interface-name-prefix': 0,
    '@typescript-eslint/no-implied-eval': 0,
    '@typescript-eslint/no-non-null-assertion': 0,
    '@typescript-eslint/no-throw-literal': 0,
    '@typescript-eslint/no-use-before-define': 0,
    '@typescript-eslint/no-var-requires': 0,
    'eslint-comments/no-unlimited-disable': 1,
    'react-hooks/exhaustive-deps': 0,
    'react-hooks/rules-of-hooks': 2,

    /**
     * ---------圈复杂度-----------
     */
    complexity: [
        'warn',
        { max: 20 }
    ],

    /**
     * ---------JSDOC-----------
     */
    // 检测无效的填充块
    'jsdoc/check-indentation': 1,
    // *号必须按标准对齐
    "jsdoc/check-alignment": 1,
    // 检测example中的js代码是否符合eslint规则
    "jsdoc/check-examples": 1,
    // 检测无效的填充块
    "jsdoc/check-indentation": 1,
    // 检测param的名称是否与函数中的参数名称一致
    "jsdoc/check-param-names": 1,
    "jsdoc/check-syntax": 1,
    // 检测块标记名称是否正确，例如@paramsss是错误的
    "jsdoc/check-tag-names": 1,
    // 检测参数类型是否正确，例如{object},可自动修复大小写
    "jsdoc/check-types": 1,
    "jsdoc/implements-on-classes": 1,
    // 通过正则表达式限制描述的内容
    // 比如限制只能使用英文，首字母大写等场景
    "jsdoc/match-description": 1,
    // 强制描述的内容后面必须有一个空行，也可以配置没有
    "jsdoc/newline-after-description": 1,
    "jsdoc/no-types": 1,
    // 不允许出现未定义的参数类型，例如{strrrrr}
    "jsdoc/no-undefined-types": 1,
    // 要求所有函数都要有描述
    "jsdoc/require-description": 1,
    // 要求完整的标记，例如描述必须使用@description，而不是直接文字描述
    "jsdoc/require-description-complete-sentence": 1,
    // 要求所有函数必须要有example
    "jsdoc/require-example": 1,
    "jsdoc/require-hyphen-before-param-description": 1,
    // 检查类声明和函数上是否有jsdoc注释
    "jsdoc/require-jsdoc": 1,
    // 要求函数所有参数写param
    "jsdoc/require-param": 1,
    "jsdoc/require-param-description": 1,
    // 要求每个@param标签都要有description
    "jsdoc/require-param-name": 1,
    // 要求每一个@param标签都要有type类型参数
    "jsdoc/require-param-type": 1,
    "jsdoc/require-returns": 1,
    // 当使用@returns后，检测函数中是否有符合条件的返回值
    "jsdoc/require-returns-check": 1,
    // @returns必须要有description
    "jsdoc/require-returns-description": 1,
    // @returns必须要有返回类型参数
    "jsdoc/require-returns-type": 1,
    // 验证类型的有效性
    "jsdoc/valid-types": 1
  },
};

const { LINT_CONFIG } = process.env

module.exports = LINT_CONFIG === 'STRICT_MODE' ? STRICT_MODE : NORMAL_MODE;
